### ISSUE TITLE
Use case: Response for Result<Something, StatusCode>

### ISSUE THREAD

**Issue by vorner** [carllerche/tower-web#86]
Hello

I was thinking that being able to do something like this would be comfortable:

```rust
#[get("/something/:path")]
fn something(&self, path: String) -> Result<String, StatusCode> {
  if &path != "stuff" {
    return Err(StatusCode::NOT_FOUND);
  }
  return Ok(path);
}
```

Does that feature make sense?

I'm not sure if it is possible now (but the documentation doesn't suggest it is possible) and if not, how to get there ‒ so I'd like to get some suggestions how to go about writing a PR.

**Comment by shepmaster** [comment]
My thoughts...

1. Investigate the `catch` handler, but it looks like it doesn't accept arguments at the moment:

    https://github.com/carllerche/tower-web/blob/533da53d7e429492d93ad550ac2c62507b6a5488/tower-web-macros/src/resource/catch.rs#L16-L17

    Adding that support seems like it would be welcomed, based on the comments.    

1. A higher-order `Service` that takes the result of the inner service and modifies it. You can see an example of that with a [WIP cache header implementation I have](https://github.com/carllerche/tower-web/issues/82#issuecomment-416058258).

1. Some middleware that does the same thing?

**Comment by carllerche** [comment]
Going from arbitrary error type -> Response without specialization is going to be hard.

the handler error types get converted into a tower-web::Error before the catch handler is called. So, if you impl a catch handler, you will end up only being able to see an opaque `tower_web::Error`.

If we had specialization, we could have an impl From<T> for tower_web::Error and then have more specialized impls.

**Comment by shepmaster** [comment]
It could be that the `Error` of a handler should always be `tower_web::Error` and represent "real hard failures". Then this instead this could be:

```
fn something(&self, path: String) -> Result<Result<String, StatusCode>, tower_web::Error>
```

Or perhaps

```rust
fn something(&self, path: String) -> Either<String, StatusCode>
```

**Comment by vorner** [comment]
> `fn something(&self, path: String) -> Result<Result<String, StatusCode>, tower_web::Error>`

That could be an option, yes. And I probably would know where to start attacking that, if you agree. Something like:


```rust
impl Response for StatusCode { ... }

impl<T: Response, E: Response> for Result<T, E> { ... }
```

And probably delegating the `Result` into `Either` (I've glimpsed some merge request around that, but I'm not sure it is exactly what I need ‒ I'd have to have a look).

Basically, I want to avoid something like this:

```rust
#[derive(Resolve)]
#[serde(untagged)]
enum WebResult {
  Page(String),
  Error {
    #[web(status)]
    status: StatusCode,
  }
}
```

**Comment by carllerche** [comment]
Your strategy looks good.

One question: Would you expect the provided status code to go through the catch handler? What will the response body be? Does only a `StatusCode` have enough info to be converted into a full response?

This also relates to #89

**Comment by vorner** [comment]
No, I wouldn't. I understand the catch handler to be something like „Something went terribly wrong“, which eg. 404 not found doesn't seem to fit (maybe I understand the intention there wrong).

The body would probably be just `text/plain` with the user-visible message („Not found“), like in eg. apache if you don't configure your own 404 page. So yes, it would be enough info to generate a full response.

**Comment by vorner** [comment]
Going through the documentation, all the examples of the handlers have `()` as an error. Should the examples be using (or at least some of them) `tower_web::Error`?

**Comment by carllerche** [comment]
@vorner re: `()` error... good question! I don't know. Arbitrary errors are converted to InternalError.

Re: catch handler, handling 404 is also an intended use case. It is there to allow customizing the response.

My guess is that it should be bound by something... maybe `Any` so that it can be passed into the catch handler.

That said, `StatusCode` could still impl `Response` and basically be an "easy way" to do something if you don't care about the details of the response. Clearly, if you did, you could always set the error to some other `T: Response` type.

**Comment by vorner** [comment]
Hmm, I think an example of how to use these errors properly would be nice, then.

I mean, if I have something like `/users/:username` path, it matches and then I have to look into database and if I don't find the user, generate a 404. This 404 isn't auto-generated. And I can't say I see the recommended way to do it manually right now.

**Comment by lnicola** [comment]
@vorner Right now I'm using `hyper::Response` directly and return `Result<Response, Response>` in situations like yours. You could also do with enums and `#[web(either)]`. But you said you'd rather avoid this, can you go a little more into why?

I think implementing `Response` for `StatusCode` isn't _that_ useful, because often (if you're not writing a web service, at least) you want to include a response body that looks nice.

**Comment by manifest** [comment]
@lnicola Could you please show an example of using `Result<Response, Response>` in resource handler.

**Comment by lnicola** [comment]
@manifest Sorry for answering so late. I have a couple of canned responses for various errors (Unauthorized, Internal Server Error, redirects and so on), which I can stick in an `Err` if needed.

**Comment by carllerche** [comment]
@lnicola Are these canned responses something that you can handle with a catch handler?

My gut is that we should have action handlers have `Error: Into<tower_web::Error>`. This would mean that you can't define http response on error in the handler, you would have to do that in the catch handler.

**Comment by lnicola** [comment]
@carllerche are catch handlers documented? I can't find anything about them.

But if they work [like this](https://github.com/carllerche/tower-web/blob/master/examples/catch.rs#L21), I'm not sure how I could pass state from the route to the catch handler.

I could presumably use `Either` or middlewares instead, so `Result<Response, Response>` isn't necessary. And I just notice I'm not actually using it any more.

**Comment by carllerche** [comment]
Looks like they aren't documented 👎 Also, the impl is not complete. The error is not passed to the handler yet :( (probably why it isn't documented yet).

Example of what works today: https://github.com/carllerche/tower-web/blob/master/examples/catch.rs
